File: Java/Platform/Basics/AsyncParallelTest/Program2.java
import java.util.stream.IntStream;
import java.util.concurrent.CompletableFuture;

class Program {

	public static void main(String[] args) throws Exception {
		int limit = Integer.parseInt(args[0]);
		System.out.print("Computing...");
		var c = new Computation();
		var t1 = System.nanoTime();
		var job = c.computeAsync(1, limit)
				   .thenAccept(r -> {
						var t2 = System.nanoTime();
						System.out.println("Done!");
						System.out.printf("Result = %d, computed in %.3f seconds.%n", r, (t2 - t1) / 1e9);
					});
		while(!job.isDone()) {
			System.out.print('.');
			Thread.sleep(500);
		}
	}

	static class Computation {

		public long compute(int first, int last) {
			return IntStream.range(first, last + 1)
					.mapToLong(Worker::doWork)
					.sum();
		}

		public CompletableFuture<Long> computeAsync(int first, int last) {
			//the supplied method is executed by a pooled thread allowing the
			//caller thread to resume its execution and obtain the result of
			//invocation at a time in future when the poooled thread has 
			//completed the execution
			return CompletableFuture.supplyAsync(() -> compute(first, last));
		}
	}
}

File: dotnet/Platform/Basics/AsyncParallelTest/DemoApp/Program2.cs
class Program
{
    private static Task DoComputation(int limit)
    {
        Console.Write("Computing...");
        var c = new Computation();
        var w = new System.Diagnostics.Stopwatch();
        w.Start();
        return c.ComputeAsync(1, limit)
                .ContinueWith(previous => 
                {
                    long r = previous.Result;
                    w.Stop();
                    Console.WriteLine("Done!");
                    Console.WriteLine("Result = {0}, computed in {1:0.000} seconds.", r, w.Elapsed.TotalSeconds);
                });
    }

    static void Main(string[] args)
    {
        int n = int.Parse(args[0]);
        var job = DoComputation(n);
        while(!job.IsCompleted)
        {
            Console.Write(".");
            Thread.Sleep(500);
        }
    }

    class Computation
    {
        public long Compute(int first, int last)
        {
            return Enumerable.Range(first, last - first + 1)
                        .Select(Worker.DoWork)
                        .Sum();
        }

        public Task<long> ComputeAsync(int first, int last)
        {
            //the operation passed to Run is invoked by a pooled
            //thread allowing the caller thread to resume and obtain
            //the result of invocation from the retuned Task once the
            //pooled thread completes that invocation
            return Task<long>.Run(() => Compute(first, last));
        }
    }
}